src/ghci/mod.rs

```diff
@@ -356,6 +424,40 @@ impl Ghci {
     }
 }
 
+/// The mode a `ghci` session is in. This is used to track output, particularly for the error log
+/// (`ghcid.txt`).
+///
+/// Note: The [`Ord`] implementation on this type determines the order in which sections will be
+/// written to the error log.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Mode {
```

Comment:
New concept! The `GhciStdin` task keeps track of the mode (because the input it sends over stdin is what causes `ghci` to do different things...) and sends messages to the other tasks when the mode needs to be updated.